Source/Core/Gen5/Generators/IDGenerator5.cpp

```diff
@@ -50,19 +49,13 @@ std::vector<IDState5> IDGenerator5::generate(u64 seed, u32 pid, bool checkPID)
 
         if (filter.compare(state))
         {
-            if (checkPID)
-            {
-                bool idBit = (tid & 1) ^ (sid & 1);
+            bool idBit = (tid & 1) ^ (sid & 1);
 
-                // Check if PID will be modified by the tid/sid combo
-                u16 actualPSV = (idBit ^ pidBit) ? xorPSV : psv;
-                if ((actualPSV >> 3) == state.getTSV())
-                {
-                    state.setSeed(seed);
-                    states.emplace_back(state);
-                }
-            }
-            else
+            // Check if PID will be modified by TID/SID combo if XOR box is checked
+            bool isShinyXor = checkXOR && (idBit == pidBit) && (psv >> 3) == state.getTSV();
```

Comment:
I don't think this check is correct. See above comment about removing `xorPSV`. You need to check which PSV you are using depending on the result of (idBit == pidBit).

I think something like this would be correct for this block. Very verbose but it should give the right idea.
```
if (filter.compare(state))
{
    if (checkPID)
    {
        if (checkXOR)
        {
            bool idBit = (tid & 1) ^ (sid & 1);
            u16 actualPSV = (idBit ^ pidBit) ? xorPSV : psv;
            if ((actualPSV >> 3) == state.getTSV())
            {
                state.setSeed(seed);
                states.emplace_back(state);
            }
        }
        else
        {
            if ((psv >> 3) == state.getTSV())
            {
                state.setSeed(seed);
                states.emplace_back(state);
            }
        }
    }
    else
    {
        state.setSeed(seed);
        states.emplace_back(state);
    }
}
```

Reply:
What are thoughts on simplifying it to remove nesting and make the code a bit more DRY:
```cpp
if (filter.compare(state)) {
    bool idBit = (tid & 1) ^ (sid & 1);

    // Check if PID will be modified by TID/SID combo if XOR box is checked
    u16 psvToTest = !checkXOR || (checkXOR && idBit == pidBit) ? psv : xorPSV;
    bool isShiny = (psvToTest >> 3) == state.getTSV();

    if (!checkPID || isShiny)
    {
        state.setSeed(seed);
        states.emplace_back(state);
    }
}
```

Reply:
> I don't think this check is correct. See above comment about removing `xorPSV`. You need to check which PSV you are using depending on the result of (idBit == pidBit).
> 
> I think something like this would be correct for this block. Very verbose but it should give the right idea.
> 
> ```
> if (filter.compare(state))
> {
>     if (checkPID)
>     {
>         if (checkXOR)
>         {
>             bool idBit = (tid & 1) ^ (sid & 1);
>             u16 actualPSV = (idBit ^ pidBit) ? xorPSV : psv;
>             if ((actualPSV >> 3) == state.getTSV())
>             {
>                 state.setSeed(seed);
>                 states.emplace_back(state);
>             }
>         }
>         else
>         {
>             if ((psv >> 3) == state.getTSV())
>             {
>                 state.setSeed(seed);
>                 states.emplace_back(state);
>             }
>         }
>     }
>     else
>     {
>         state.setSeed(seed);
>         states.emplace_back(state);
>     }
> }
> ```

We don't need this check. We need IDs combo for our wanted PID, not for the xored one.
If you leave the check, you'll get IDs combo to make the wanted PID shiny, but you'll get also IDs combot that will make the PID^0x80000 to be shiny. We don't want the last thing

Reply:
Yeah looks good zak

Reply:
If we check (psv ^ 0x8000), the user will always be able to get a shiny but in the half of the cases he will not get the specified pid.

Maybe something like this would be fine:
```c++
if (filter.compare(state)) {
    
    bool isShiny = (psv >> 3) == state.getTSV();

    // Check if PID is possible with TID/SID combo if Wild/Stationary box is checked
    if (checkXOR)
        isShiny &= ((tid & 1) ^ (sid & 1)) == pidBit;
    
    if (!checkPID || isShiny)
    {
        state.setSeed(seed);
        states.emplace_back(state);
    }
}
````

Reply:
We should be able to replace the `&=` with `=` to be a bit more readable:
```cpp
isShiny = ((tid & 1) ^ (sid & 1)) == pidBit;
```

Due to the if statement, the left hand side of that calculation will always be `true`.

Reply:
The above change has been done